Fix Reporting Bugs

timburke opened the pull request:
This is a work in progress intended for collaboration.

There are also (optional) changes to the pic12 executive included that allow for bus resetting when the gsm_module is enabled.

timburke committed:
Add 10 us low alarm pulse resets all slaves.

- Reimplement bus slave logic in assembly for optimized size
- Changed several functions to indicate return value using status
bits like ZERO for optimized checks.
- Implemented bus_slave_callhandler in assembly

timburke committed:
Update test logic to properly test ASM version of bus logic

timburke committed:
Add peripheral module for Interrupt on Change functionality

timburke committed:
Properly clear IOC flag after checking

timburke committed:
Disable IOC when we reflash so that we don't reset ourselves.

timburke committed:
Remove legacy C versions of assembly file.

timburke committed:
Remove duplicate status message

timburke committed:
Increase pic12_executive serial number to reflect IOC addition

timburke committed:
Update gsm module to use new pic12_executive API

timburke committed:
Remove default reporting call

timburke committed:
Remove alarm check for running application code

Upon ALARM assertion, all modules will reset automatically based
on the IOC interrupt implemented on the ALARM pins, so there is
no possibility of running application code (or any code) while the
alarm pin is down.

timburke committed:
Implement reset delay to allow modules to reset themselves

Adds a 100 ms wait from the time ALARM is asserted to the time the
bus power goes off to allow chips to reset and clear all of their
pins.  This disables any additional circuitry on the bus that may
be power the bus parasitically when main power is cut and prevent
a proper reset sequence.

timburke committed:
Fix RTCC to use internal RC oscillalator

The external crystal was not reliably working, leading to the RTCC
being enabled but not triggering scheduled callbacks because it
did not receive ticks from the crystal.

timburke committed:
Fix modtool time command to use 1-base month and day

This is what the RTCC on the PIC24 reports.

timburke committed:
Make rtcc not fill up task loop during long running tasks

This is a bug fix.  If there is a periodically scheduled task
with interval N and a task takes more than taskloop_size*N seconds,
the rtcc_alarm will add the_alarm_callback 32 times to the taskloop
without it ever running.  This fixes that by not adding a task to
the taskloop if the previous addition has not yet finished runing.

timburke committed:
Redo timestamp generation to be more efficient

New implementation uses logarithmic lookup tables to avoid any
32 bit multiplications.  Implementation complexity is log(N) in the
number of bits required to represent the datetime and only uses
32 bit addition with a few small lookup tables.

timburke committed:
Change sensor logging to use uint32 timestamps

This replaces using HH:MM:SS DD/MM/YY timestamps since time
delta compuatation is much faster with integral timestamps and there
is no loss of precision.

timburke committed:
Reimplement timestamp generation and comparison

- Revert my previous implementation using lookup tables since 16x16
multiplication on pic24 is single cycle so there is no reason to
avoid it.

timburke committed:
Reimplement construct report to avoid pointer math

timburke committed:
Revert logarithmic table lookup routine

timburke commented:
@amcgee This is the latest  branch that I've been working on.

amcgee commented:
Great, thanks.  I'll take a look tonight.
On Jun 23, 2014 7:23 PM, "timburke" notifications@github.com wrote:

> @amcgee https://github.com/amcgee This is the latest branch that I've
> been working on.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/WellDone/MoMo-Firmware/pull/103#issuecomment-46867506
> .

timburke committed:
Update system log to use uint32 timestamps

timburke committed:
Add lightweight performance counter module

Module configures a freerunning 32-bit timer running at the same
frequency as the CPU instructions.  An arbitray number of perormance
counters can be configured to micro-profile individual regions of code.

Calling PROFILE_START(counter_id) captures the current timer value.
Calling PROFILE_END(counter_id) stores the elapsed time since the
last call to PROFILE_START with the counter id.  The times can be
recalled later using perf_get_counter.

I used this to fix a huge flash memory performance problem.

timburke committed:
Add profiling and fix giant performance problem

Due to the way the SPI peripheral was initialized, it was not running
at full speed.  Evidently, you must set the SPI clock frequency
before setting the SPIEN bit to enable the module.  Failing to do
this causes the setting to be ignored.  This manifests in reading and
writing from flash taking between 3 and 10 times longer than necessary
and causing performance issues in construct_report as well as making
modtool reflash run very slowly.

Using the new perf module, I tracked this down and fixed it.

timburke committed:
Add support for reading performance counters and arbitrary RAM

- adds API interface for reading perf counters
- adds non-negative validator for integer types in typedarg
- adds API interface for reading arbitrary RAM locations for debugging

timburke committed:
Add support for new report format in reportinator tool

- Uses uint32 timestamps rather than 6-byte datetime codes

timburke committed:
Add support for reading uint32 timestamps in system_log

- Needs to be extended still to compute the proper date and time
from the timestamp (timestamp stores elapsed time since midnight
1/1/1 2000)

timburke committed:
Add MIB command for building a report on demand and dumping it
over the bus.

timburke committed:
BUGFIX: off by one error in firmware cache

Would allow 8-bit firmware to overwrite bucket 4,which is controller
firmware only.  Should not have caused any problems but it's incorrect
and simple to fix.  modtool reflash would have overwritten the bad
data before reflashing the controller anyway.

timburke committed:
Add support for perf counters and reading ram

- also drive bus voltage transistor rather than just letting it
get pulled up to VCC to make it more robust against noise.  I do
not believe that this caused a problem before but it's better to
drive signals than to let them be pulled up by high value resistors

timburke commented:
Update on this PR.
- I fixed a major bug in memory.c that caused the flash reading to be 3-10 times slower than it needed to be.  This could make construct report take an extremely long time to return (>5 minutes).  With a large number of events to process (~1000) construct_report can still take over 30 seconds.  
- Added a new perf module to do lightweight profiling of arbitrary code regions.  I used this to track down the flash memory issue.
- Converted tools to use timestamps rather than datetime for handling times since the format is more compact and more amenable to fast processing.

TODO:
- timestamp generation is not correct I don't think.  I don't see reports in individual buckets when I construct a report an then dump it.  I do see the correct number of reports in the bulk aggregate counter.  I think there's a bug in the timestamp generation routine in rtccc.c
- construct_report is still very slow though no longer crashes the PIC.  Profiling is needed to figure out where the slowness comes from.  The new perf module can be used for this.

timburke committed:
Update peripheral pin mapping to map SCK as IN/OUT

Per the datasheet on the chip, you must map SCK as both an input
and an output.  I don't think this has caused any issues so far
but it's better to comply with the spec than not to.

timburke committed:
Update performance counting module to keep min, max, count and avg

Average is an exponentially filtered running average over an 8
element window so it should converge to the average runtime of
the last 8 or so profiling event.

Support for this has been added to the momo tool.  Just use
```shell
> momo controller perf_counter <NUMBER>
```

timburke committed:
Fix interval calculation bug in construct_report

construct_report function should now be safe and working.  Command
line testing shows it to function correctly.  It can process about
300 events per second, so large event queues will still be slow.

Major slowdown now seems to be in reading the sensor events in from
flash.  This is currently done 1 at a time so increasing the read
buffer will probably help.

timburke commented:
This PR should now fix the issues with construct_report.  It still can only process ~300 events per second, but it appears to be stable in testing and the constructed reports seem correct.  Report construction testing is now possible using:

``` shell
$ momo controller build_report
$ momo controller get_report | xargs reportinator parse
version: 2
sensor: 11
sequence: 49
flags: 0
battery charge: 3.84V
diag: 431, 0
Interval type: second
         step: 6
        count: 10
Global aggregates:
    count = 431
    min = 15
    max = 15
Interval Aggregates:
           count    mean
               6      15
               6      15
               6      15
               6      15
               6      15
               6      15
               6      15
               6      15
               6      15
               5      15
```

Relevant performance counters are:
- (5) total time to construct the report
- (6) total time to encode the report in base 64 (included in 5)
- (7) time to process a single event

These counters can be read by using, e.g.

``` shell
$ momo controller perf_counter 7
Performance Counter (7)
Number of Events: 680
Min: 19.00 us (76 cycles)
Max: 72.80 ms (291193 cycles)
Mean: 2.62 ms (10478 cycles)
```

timburke commented:
Note that the perf_counter mean is not valid when there are fewer than 8 events.  It is an exponential running average with a window of 8, so it will just be artificially low when there are fewer than 8 samples to average.

amcgee commented on momo_modules/mainboard/src/mib/controller_mib_feature.c:
```diff
@@ -277,6 +286,23 @@ void set_lazy_logging()
+void get_perf_counter()
+{
+	uint16 counter = plist_get_int16(0);
+	uint32 val = perf_get_counter((PerformanceCounter)counter);
```
This line causes a compiler warning:

```
src/mib/controller_mib_feature.c:298:15: warning: initialization makes integer from pointer without a cast
```

I think it should actually pass a 20-byte buffer in order to work with the python code, otherwise I don't know how this could work.

Does that seem right?  I'll push the changes when I have a chance to test them.

timburke replied:
I must not have committed that last change.  Sorry.  Yeah it passes a 
buffer.  If you mean the get_perf_counter function.  This is an earlier 
version that just had one number for each perf counter.  A later commit 
should have upped it to 5 uint32s and returned it as a buffer.
- tim

On 6/27/14, 6:55 AM, Austin McGee wrote:

> In momo_modules/mainboard/src/mib/controller_mib_feature.c:
> 
> > @@ -277,6 +286,23 @@ void set_lazy_logging()
> >     lazy_system_logging = (plist_get_int16(0)==0)?false:true;
> >  }
> > 
> > +void read_ram()
> > +{
> > -   unsigned char *val = (unsigned char *)(plist_get_int16(0));
> > -   bus_slave_return_buffer(val, 20);
> >   +}
> >   +
> >   +void get_perf_counter()
> >   +{
> > -   uint16 counter = plist_get_int16(0);
> > -   uint32 val = perf_get_counter((PerformanceCounter)counter);
> 
> This line causes a compiler warning:
> 
> |src/mib/controller_mib_feature.c:298:15: warning: initialization makes integer from pointer without a cast
> |
> 
> I think it should actually pass a 20-byte buffer in order to work with 
> the python code, otherwise I don't know how this could work.
> 
> Does that seem right? I'll push the changes when I have a chance to 
> test them.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/WellDone/MoMo-Firmware/pull/103/files#r14293330.

amcgee commented:
This looks good and has worked well with minimal testing.  More to come this evening / weekend.

I'm not 100% convinced this is the issue seen in Iringa, the symptoms were almost always a (relatively) short construct_report call followed by some writing to a stream and then a hang (open, write, write, hang....)  This could be related to the memory corruption in the task_loop data structures, is that the theory?

In testing I've had reasonable functionality when network signal is strong, but unpredictable behavior when it's weak (a side-effect of which is really long open_stream calls hanging the bus).  As mentioned, testing more this weekend.

We should avoid global static performance counter declarations if possible, but it works here.  We could also create a performance log that's optionally enabled.

timburke commented:
This is probably Issue #102.  The GSM transmission, when it is really 
powerful causes the voltage regulator powering the PICs to cut out.  
This is an RF interference problem and apparently internal circuitry 
inside the regulator is susceptible to the interference. I debugged it 
with an oscilloscope.

The fix is to wrap the entire momo in paper (or something nonconductive 
and then wrap it all in aluminum foil.  The foil forms an RF shield.  
This fixes the issue in my testing if the foil is continuous around the 
device.  So, I would assemble the MoMo units, connect the antenna wrap 
them in the paper+foil and make sure the antenna sticks out of the foil 
so it's on the outside.  You can have the MoMo debug cable to the 
computer through the foil as well.  When I do this it fixes the problem.

Could you try that and see if it helps?
- tim

On 6/27/14, 6:59 AM, Austin McGee wrote:

> This looks good and has worked well with minimal testing. More to come 
> this evening / weekend.
> 
> I'm not 100% convinced this is the issue seen in Iringa, the symptoms 
> were almost always a (relatively) short construct_report call followed 
> by some writing to a stream and then a hang (open, write, write, 
> hang....) This could be related to the memory corruption in the 
> task_loop data structures, is that the theory?
> 
> In testing I've had reasonable functionality when network signal is 
> strong, but unpredictable behavior when it's weak (a side-effect of 
> which is really long open_stream calls hanging the bus). As mentioned, 
> testing more this weekend.
> 
> We should avoid global static performance counter declarations if 
> possible, but it works here. We could also create a performance log 
> that's optionally enabled.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/WellDone/MoMo-Firmware/pull/103#issuecomment-47347204.

timburke committed:
Change controller mib feature to use new perf counter format.
